### PR TITLE
feat: add cross device sync

### PR DESCRIPTION
**Pull request by kaiserbh**
This essentially refers to the pull request I submitted at [GitHub - Tachiyomi Pull Request #9790](https://github.com/tachiyomiorg/tachiyomi/pull/9790). My initial thought was to contribute the pull request to the upstream project, anticipating that it would subsequently be integrated into the downstream, particularly SY, which I predominantly use. However, given the halt in development for the vanilla version, I opted to directly submit the pull request to SY. This PR encompasses numerous modifications; the feature in question has been available for a considerable duration and has undergone thorough testing on the Discord server, with all functionalities operating smoothly.

Currently, there are two services available: [Syncyomi](https://github.com/SyncYomi/SyncYomi), which is self-hosted, and Google Drive.

Once merged you will need to create OAuth 2.0 Client IDs in google console then replace the `app/src/main/assets/client_secrets.json`

image

Once all of that is done just need to set the app to `publish` instead of `testing` that way everyone can access it otherwise limited to 100 users we add for testing, we don't need to submit the app for approval since we are not using sensitive scopes. More info [here](https://support.google.com/cloud/answer/13463073?authuser=1&visit_id=638407747036157208-1875300124&rd=1)

### OAuth Consent Screen
image

### Scopes
image

image
image
image
image
image

### CODE REVIEW

**Comment by KaiserBh** [comment]
Opened a PR to upstream https://github.com/mihonapp/mihon/pull/40 so feel free to wait for that to be merged before merging this.

**Comment by jobobby04** [comment]
Great! Hopefully it's merged soon

**Comment by NeKoOuO** [comment]
@KaiserBh omg I just saw an issue from seven years ago talking about the need for cloud sync. I didn't expect it to actually exist now. Thank you for your great PR

**Comment by kaiserbh** [comment]
Hi @jobobby04, I have closed the PR to the upstream I explained it on the discord announcement for the people who are testing it or been using it for the past year ish, anyway, I know the PR is big and I am more than happy to streamline it and divide it to make it smaller. So whenever you got the time can you do initial review of it and let me know the changes I need to make?

File: app/src/main/java/eu/kanade/tachiyomi/App.kt
Comment (jobobby04) [inline app/src/main/java/eu/kanade/tachiyomi/App.kt:170]:
```suggestion
        val syncPreferences: SyncPreferences = Injekt.get()
```

File: app/src/main/java/eu/kanade/tachiyomi/App.kt
Comment (jobobby04) [inline app/src/main/java/eu/kanade/tachiyomi/App.kt:214]:
```suggestion
        val syncPreferences: SyncPreferences = Injekt.get()
```

File: app/src/main/java/eu/kanade/tachiyomi/App.kt
Comment (jobobby04) [inline app/src/main/java/eu/kanade/tachiyomi/App.kt:254]:
```suggestion
        val syncPreferences: SyncPreferences = Injekt.get()
```

File: app/src/main/java/eu/kanade/tachiyomi/data/backup/models/BackupChapter.kt
Comment (jobobby04) [inline app/src/main/java/eu/kanade/tachiyomi/data/backup/models/BackupChapter.kt:25]: Can these backup numbers be PRd to Mihon? I don't want to create accidental conflicts

File: app/src/main/java/eu/kanade/tachiyomi/data/backup/models/BackupManga.kt
Comment (jobobby04) [inline app/src/main/java/eu/kanade/tachiyomi/data/backup/models/BackupManga.kt:44]: Same with these

File: app/src/main/java/eu/kanade/tachiyomi/data/sync/SyncManager.kt
Comment (jobobby04) [inline app/src/main/java/eu/kanade/tachiyomi/data/sync/SyncManager.kt:190]:
```suggestion
            cacheFile.outputStream().use { output ->
```

File: app/src/main/java/eu/kanade/tachiyomi/data/sync/service/GoogleDriveSyncService.kt
Comment (jobobby04) [inline app/src/main/java/eu/kanade/tachiyomi/data/sync/service/GoogleDriveSyncService.kt:178]:
```suggestion
            GZIPOutputStream(byteArrayOutputStream).use { gzipOutputStream ->
                gzipOutputStream.write(jsonData.toByteArray(Charsets.UTF_8))
            }
```

**Review by jobobby04 — COMMENTED** [review]
Code looks good, some nitpicks here and there

File: app/src/main/java/eu/kanade/tachiyomi/data/backup/models/BackupChapter.kt
Comment (kaiserbh) [reply app/src/main/java/eu/kanade/tachiyomi/data/backup/models/BackupChapter.kt:25]: Done opened PR:   https://github.com/mihonapp/mihon/pull/450

**Comment by jobobby04** [comment]
I've pushed the latest Mihon commits, looks like conflicts need fixing

**Comment by kaiserbh** [comment]
> I've pushed the latest Mihon commits, looks like conflicts need fixing

fixed.

**Review by jobobby04 — APPROVED** [review]

**Pull request merged**
